feat: change the default data storage version to "stable" (e.g. v2.0)

westonpace opened the pull request:
Closes #2394

wjones127 reviewed (commented):
We should probably state the earliest version of Lance that can read these new V2 files. I think we are breaking a certain backwards compatibility by changing the default, right?

westonpace commented:
> I think we are breaking a certain backwards compatibility by changing the default, right?

Backwards compatibility, no, I don't think so.  We won't lose any capability to read older files.

Forwards compatibility: kind of.

This only applies the v2 default to _new_ datasets.  So old readers will continue to be able to read old datasets but yes, readers beyond a certain point (probably 0.16 to be safe but in reality maybe even older) will not be able to read new datasets created by new versions.  I'll update some docs somewhere.

westonpace commented:
> We should probably state the earliest version of Lance that can read these new V2 files. I think we are breaking a certain backwards compatibility by changing the default, right?

I've marked this a breaking change.  I verified that we have a "minimum lance version" in the format section (and updated it from 0.15 to 0.16).  I've added some basic "future compatibility" tests to make sure we can, in fact, read lance files written with the latest version in 0.16.

wjones127 commented on .github/workflows/python.yml:
```diff
@@ -105,6 +105,17 @@ jobs:
+        python -m venv venv
+        source venv/bin/activate
+        pip install pytest pylance==0.16.0
+    - name: Run forward compatibility tests
+      run: |
+        source venv/bin/activate
+        pytest python/tests/forward_compat --run-forward
```
Fancy! ✨

wjones127 commented on python/python/tests/test_dataset.py:
```diff
@@ -538,6 +539,31 @@ def test_pickle(tmp_path: Path):
+
+    dataset = lance.dataset(base_dir)
+
+    print(format_fragment(dataset.get_fragment(0).metadata, dataset))
```
Should we be keeping this print?

wjones127 commented on rust/lance/src/io/exec/pushdown_scan.rs:
BTW, do we support pushdown in V2 yet? Is that coming soon?

westonpace replied:
No.  Soon?  I'll try and work on that next once I'm done with this mini-block stuff.
